### ISSUE TITLE
Unable to restore-cluster where topology does not match

### ISSUE THREAD

**Issue by jfharden** [thelastpickle/cassandra-medusa#46]
I am trying to restore a cluster into a new cluster where the topology doesn't match, but I get a python exception:

```
# medusa restore-cluster --host-list /etc/medusa/restore-mapping.txt --backup-name=test001 --use-sstableloader --temp-dir /mnt/cassandra/data01/__MEDUSA_TMP_DIR/
[2020-01-07 14:56:09,577] INFO: Monitoring provider is noop
[2020-01-07 14:56:09,577] INFO: system_auth keyspace will be left untouched on the target nodes
[2020-01-07 14:56:10,383] INFO: Ensuring the backup is found and is complete
[2020-01-07 14:56:10,406] INFO: Restore will happen on new hardware
[2020-01-07 14:56:10,406] INFO: Starting Restore on all the nodes in this list: /etc/medusa/restore-mapping.txt
[2020-01-07 14:56:10,406] INFO: Starting cluster restore...
[2020-01-07 14:56:10,406] INFO: Working directory for this execution: /mnt/cassandra/data01/__MEDUSA_TMP_DIR/medusa-job-a27776bf-eb6e-4950-99b3-d619aa8a1a38
[2020-01-07 14:56:10,406] INFO: About to restore on 172.17.41.10 using {'source': ['ip-172-17-44-6.eu-west-1.compute.internal'], 'seed': True} as backup source
[2020-01-07 14:56:10,406] INFO: This will delete all data on the target nodes and replace it with backup test001.
Are you sure you want to proceed? (Y/n)Y
[2020-01-07 14:56:11,891] INFO: target seeds : ['172.17.41.10']
[2020-01-07 14:56:11,891] INFO: Restoring schema on the target cluster
[2020-01-07 14:56:11,934] ERROR: This error happened during the cluster restore: 'NoneType' object has no attribute 'new_session'
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/medusa/restore_cluster.py", line 72, in orchestrate
    restore.execute()
  File "/usr/local/lib/python3.7/site-packages/medusa/restore_cluster.py", line 147, in execute
    self._restore_data()
  File "/usr/local/lib/python3.7/site-packages/medusa/restore_cluster.py", line 328, in _restore_data
    self._restore_schema()
  File "/usr/local/lib/python3.7/site-packages/medusa/restore_cluster.py", line 384, in _restore_schema
    with self.session_provider.new_session() as session:
AttributeError: 'NoneType' object has no attribute 'new_session'
```

As far as I can see this code path can't work. The session_provider is only set in the if statement where you have given a seed provider:

https://github.com/thelastpickle/cassandra-medusa/blob/master/medusa/restore_cluster.py#L130-L144

if you have provided a host list instead you will hit the exception above, providing both is an error: 

```
# medusa restore-cluster --seed-target 172.17.41.10 --host-list /etc/medusa/restore-mapping.txt --backup-name=test001 --use-sstableloader --temp-dir /mnt/cassandra/data01/__MEDUSA_TMP_DIR/
[2020-01-07 14:57:44,946] INFO: Monitoring provider is noop
[2020-01-07 14:57:44,946] ERROR: You must either provide a seed target or a list of host, not both
[2020-01-07 14:57:44,946] ERROR: This error happened during the cluster restore: You must either provide a seed target or a list of host, not both
Traceback (most recent call last):
  File "/usr/local/lib/python3.7/site-packages/medusa/restore_cluster.py", line 49, in orchestrate
    raise Exception(err_msg)
Exception: You must either provide a seed target or a list of host, not both
```

It's perfectly possible I'm doing something wrong but the docs on how to restore-cluster to an alternate topology are a little thin on the ground (I'm happy to try and help with this if I can figure out the correct incantations).

**Comment by adejanovski** [comment]
Hi,

Cross topology remote restores, if I remember correctly, should use the seed target (one active node in the cluster where data will be loaded) and no host-list. You should not even need to specify --use-sstableloader as Medusa will detect that topologies don't match and use the sstableloader right away.
Let me know if it works this way.
Happy to have some improvements in the docs once you get it working :+1:

**Comment by adejanovski** [comment]
And our logging is clearly misleading here, because we say `Restore will happen "In-Place", no new hardware is involved` which is obviously not the case 😅

**Comment by jfharden** [comment]
@adejanovski Thanks, it turns out I was also getting hit by https://github.com/thelastpickle/cassandra-medusa/issues/13

I was able to start my restore, although the temp directory is not respected on the remote machines, I had to kill the restore process since it was going to exhaust my root disk pretty quickly.

Source cluster (3 node):
172.17.40.6
172.17.42.6
172.17.44.6

Destination cluster (1 node for now, once I know it's working I'm going to try to a 6 node cluster):
172.17.41.10

They have cassandra data in /mnt/cassandra/data01.
The total backup size of the source cluster is 1.8TB, so I've mounted a 4TB disk (I can make this a lot bigger if need be) on the destination cluster into /mnt/cassandra/data01.

I ran the following restore command on 172.17.40.6:

```
[ec2-user@ip-172-17-40-6 ~]$ medusa restore-cluster --seed-target 172.17.41.10 --backup-name=test002 --temp-dir /mnt/cassandra/data01/__MEDUSA_TMP_DIR/
[2020-01-08 10:37:17,465] INFO: Monitoring provider is noop
[2020-01-08 10:37:17,465] INFO: system_auth keyspace will be overwritten with the backup on target nodes
[2020-01-08 10:37:18,264] INFO: Ensuring the backup is found and is complete
[2020-01-08 10:37:18,277] INFO: Restore will happen "In-Place", no new hardware is involved
[2020-01-08 10:37:18,298] INFO: Starting Restore on all the nodes in this list: None
[2020-01-08 10:37:18,298] INFO: Starting cluster restore...
[2020-01-08 10:37:18,298] INFO: Working directory for this execution: /mnt/cassandra/data01/__MEDUSA_TMP_DIR/medusa-job-8eb30f54-3b16-473f-9499-ac1a3c32becc
[2020-01-08 10:37:18,298] INFO: About to restore on ip-172-17-41-10.eu-west-1.compute.internal using {'source': ['ip-172-17-44-6.eu-west-1.compute.internal', 'ip-172-17-40-6.eu-west-1.compute.internal', 'ip-172-17-42-6.eu-west-1.compute.internal'], 'seed': False} as backup source
[2020-01-08 10:37:18,298] INFO: This will delete all data on the target nodes and replace it with backup test002.
Are you sure you want to proceed? (Y/n)Y
[2020-01-08 10:37:20,196] INFO: target seeds : []
[2020-01-08 10:37:20,196] INFO: Restoring schema on the target cluster
[2020-01-08 10:37:20,268] INFO: (Re)creating schema for keyspace keyspace_1
[2020-01-08 10:37:21,139] INFO: (Re)creating schema for keyspace keyspace_2
[2020-01-08 10:37:22,090] INFO: Restoring data on ip-172-17-41-10.eu-west-1.compute.internal...
[2020-01-08 10:37:22,090] INFO: Executing "nohup sh -c "mkdir /mnt/cassandra/data01/__MEDUSA_TMP_DIR/medusa-job-8eb30f54-3b16-473f-9499-ac1a3c32becc; cd /mnt/cassandra/data01/__MEDUSA_TMP_DIR/medusa-job-8eb30f54-3b16-473f-9499-ac1a3c32becc && medusa-wrapper sudo medusa --fqdn=%s -vvv restore-node --in-place  %s --no-verify --backup-name test002 --use-sstableloader  "" on all nodes.
```

On the target node it has created a temp directory in /mnt/cassandra/data01/__MEDUSA_TEMP_DIR but this only contains:

```
[root@ip-172-17-41-10 TripProfiles]# ls -la /mnt/cassandra/data01/__MEDUSA_TMP_DIR/medusa-job-8eb30f54-3b16-473f-9499-ac1a3c32becc/
total 672
drwxr-xr-x 2 root      root          79 Jan  8 10:41 .
drwxr-xr-x 5 cassandra cassandra    171 Jan  8 10:37 ..
-rw-r--r-- 1 root      root           5 Jan  8 10:37 pid
-rw-r--r-- 1 root      root           4 Jan  8 10:41 status
-rw-r--r-- 1 root      root      676896 Jan  8 10:41 stderr
-rw-r--r-- 1 root      root           0 Jan  8 10:37 stdout
```

The stderr file is showing it downloading files from s3 ok, but they are going into /tmp/medusa-restore-78a6266e-d5d3-4a7c-a409-afe2a1ed4d7c

So I note although the restore-cluster command correctly uses the temp dir in source and destination cluster the restore-node command it executes does not include the temp dir.

I've created PR

**Comment by jfharden** [comment]
So I ran into 2 more problems:

Once it was correctly downloading the files it seems to need to fit an entire file into memory before saving it to disk as part of the download, some of our sstable files are over 200Gig. It kept crashing with MemoryError:

```
[2020-01-08 12:07:12,177] DEBUG: https://s3-eu-west-1.amazonaws.com:443 "GET /bucket/ip-172-17-40-6.eu-west-1.compute.internal/test002/data/keyspace1/table1/keyspace1-table1-ka-18-Data.db?Version=2006-03-01 HTTP/1.1" 200 6865199452
Traceback (most recent call last):
  File "/bin/medusa", line 11, in <module>
    load_entry_point('cassandra-medusa==0.4.0', 'console_scripts', 'medusa')()
  File "/root/.local/lib/python3.7/site-packages/click/core.py", line 764, in __call__
    return self.main(*args, **kwargs)
  File "/root/.local/lib/python3.7/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/root/.local/lib/python3.7/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/root/.local/lib/python3.7/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/root/.local/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/root/.local/lib/python3.7/site-packages/click/decorators.py", line 64, in new_func
    return ctx.invoke(f, obj, *args, **kwargs)
  File "/root/.local/lib/python3.7/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/usr/local/lib/python3.7/site-packages/medusa/medusacli.py", line 212, in restore_node
    verify, set(keyspaces), set(tables), use_sstableloader)
  File "/usr/local/lib/python3.7/site-packages/medusa/restore_node.py", line 51, in restore_node
    keyspaces, tables)
  File "/usr/local/lib/python3.7/site-packages/medusa/restore_node.py", line 151, in restore_node_sstableloader
    download_data(config.storage, node_backup, fqtns_to_restore, destination=download_dir)
  File "/usr/local/lib/python3.7/site-packages/medusa/download.py", line 47, in download_data
    storage.storage_driver.download_blobs(srcs, dst)
  File "/usr/local/lib/python3.7/site-packages/medusa/storage/s3_storage.py", line 119, in download_blobs
    blob.download(os.path.join(blob_dest, file_name), overwrite_existing=True)
  File "/root/.local/lib/python3.7/site-packages/libcloud/storage/base.py", line 102, in download
    delete_on_failure)
  File "/root/.local/lib/python3.7/site-packages/libcloud/storage/drivers/s3.py", line 457, in download_object
    response = self.connection.request(obj_path, method='GET', raw=True)
  File "/root/.local/lib/python3.7/site-packages/libcloud/common/base.py", line 603, in request
    stream=stream)
  File "/root/.local/lib/python3.7/site-packages/libcloud/http.py", line 248, in prepared_request
    verify=self.ca_cert if self.ca_cert is not None else self.verify)
  File "/root/.local/lib/python3.7/site-packages/requests/sessions.py", line 686, in send
    r.content
  File "/root/.local/lib/python3.7/site-packages/requests/models.py", line 828, in content
    self._content = b''.join(self.iter_content(CONTENT_CHUNK_SIZE)) or b''
MemoryError
```

As a test I tried using an m5.16xlarge since it has 256GB of RAM, but the same thing happened again, so as a last ditch attempt I relaunched as an r5.24xlarge to get 768GB of ram....this actually worked.

Here is just a little debugging output to show that it really was crushing the memory:

Memory usage in megabytes (note the currently in use 175GB and the cache of 315GB)
```
              total        used        free      shared  buff/cache   available
Mem:         765724      175673      274516           0      315534      585105
Swap:             0           0           0
```

and just a little bit of the top output at the same time to show medusa using 22.3% of the available 768GB of ram
```
  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND
 8392 root      20   0  167.2g 166.8g  21344 R  54.3 22.3  50:28.20 medusa
 7719 cassand+  20   0   10.5g   8.9g 193372 S   0.7  1.2   1:08.69 java
```

Every time it crashed with a memory error it was always while downloading a particularly big sstable file. (I think our biggest is in the region of 200-250GB region).

I'm not 100% sure but i suspect this to be an issue in either the libcloud library (which I already had a problem with while testing with medusa and opened this issue: https://github.com/apache/libcloud/issues/1401) or a problem the requests library.

**Comment by jfharden** [comment]
My second problem was with the construction of the sstableloader command. It is including the cql_username and cql_password irrespective of whether they are set or not: https://github.com/thelastpickle/cassandra-medusa/blob/master/medusa/restore_node.py#L173

We don't have them set so I suspect the command ends up looking like

```
sstableloader -d 127.0.0.1 --storage-port 7000 -u --password --no-progress /path/to/sstable
```

Unfortunately the execution of the process isn't capturing or reporting on the error. My restore definitely died at the point of running this command but I can't see exactly why.

I've modified my local copy to print out some debugging info (printing out the whole command before running it so I should hopefully be able to confirm if this is indeed the problem and potentially submit a PR for it.

**Comment by adejanovski** [comment]
OK, so for problem number 1 I think we need to bypass libcloud on large files and use aws cli instead, which will trigger multi part downloads. We already do this for uploads.
It means using [download_object_as_stream()](https://github.com/apache/libcloud/blob/813b44e09245035fa4a86388fda4357509a30498/libcloud/storage/drivers/s3.py#L469) instead of blob.download() [here](https://github.com/thelastpickle/cassandra-medusa/blob/92bc8b70ad/medusa/storage/aws_s3_storage/concurrent.py#L184).
It's good to keep small files (<100MB) to use libcloud instead as there's an overhead in calling awscli for small files.

For issue number 2, we just need to put a dummy user/pass combination if it's not provided, as Cassandra is then probably using the AllowAllAuthenticator and will accept anything. I think that's what I had in mind when I coded this.

**Comment by adejanovski** [comment]
I created #52 to address the first problem and am working on a PR.

**Comment by adejanovski** [comment]
Hi @jfharden,

do still run into issues for cross topology restores? Is the sstableloader invoked as it should be despite the absence of credentials?

**Comment by liwy00** [comment]
Ran into problem that roles got lost in the restore process. Data downloaded and restore completed w/out error, but then can't log into C*

**Comment by adejanovski** [comment]
Yes, there's a `--keep-auth` flag on the restore operations to prevent Medusa from overwriting `system_auth`.

**Comment by adejanovski** [comment]
The sstableloader issue was fixed by #69.
